### PR TITLE
Fix a bug in `Bernoulli.log_prob`

### PR DESCRIPTION
**Pull request by seiyab**
Fixed a bug mentioned in https://github.com/chainer/chainer/issues/7057 .

This bug was caused by unintended broadcasting in the backward method.

I added a test that reproduces the bug and fixed it referencing this code https://github.com/chainer/chainer/blob/c92db3396d55a2a351d1bee1a47e33e85dd144b2/chainer/functions/math/basic_math.py#L176  .

### CODE REVIEW

File: chainer/distributions/bernoulli.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from chainer.backends import cuda
 from chainer import distribution
 import chainer.distributions.utils
+from chainer.functions import sum_to
```
Comment (niboshi) [inline chainer/distributions/bernoulli.py]: We do not allow importing non-module objects: https://docs.openstack.org/hacking/latest/user/hacking.html#imports

Please import like `from chainer.functions.math import sum` and use `sum.sum_to`.

**Comment by seiyab** [comment]
@niboshi 
Thank you for your review.
I fixed the import statement.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
@seiyab Could you check test failures?

**Comment by seiyab** [comment]
@niboshi 
I'll check them.

I'm sorry that I tested with only python3.
I'll reproduce them in my environment and fix them.
Now I cannot reproduce failure on my python2 environment. I might ask it when I think that I'm stuck.

Some of failures seem not to be related. (e.g. [test_convolution_2d](https://jenkins.preferred.jp/job/chainer/job/chainer_pr/699/TEST=chainer-py2,label=mn1-p100/testReport/junit/tests.chainer_tests.functions_tests.connection_tests.test_convolution_2d/TestConvolution2DFunction_use_chainerx_false__chainerx_device_None__use_cuda_true__cuda_device_0__use_cudnn_always__cudnn_deterministic_true__autotune_false__cudnn_fast_batch_normalization_false__use_ideep_never_param_31_%7Bcover_all=False,%20groups=1,%20dilate=1,%20nobias=False,%20c_contiguous=False,%20W_dtype=float32,%20x_dtype=float16%7D/test_double_backward/)
I will check causes but I might also ask it when I think that I'm stuck.

**Comment by niboshi** [comment]
@seiyab Yes, currently some of our tests occasionally fail (including convolution_2d)... Please ignore them. Sorry for inconvenience.

**Comment by seiyab** [comment]
@niboshi 
I see. I'll check test_convolution_2d a little just to make sure.

And I reproduced the future warning on my python2 environment downgrading numpy version.
I'll fix.

**Comment by seiyab** [comment]
@niboshi 
I resolved the warning in my python2 environment (python 2.7.16, numpy 1.12.1).
Of course I also tested on python3.

I have two questions.

1. Is it better that tests for infinity logits exist?
https://github.com/chainer/chainer/pull/7064/commits/a0256a672e30a6866de2e23c9370953975bf6724 changes the implementation to handle infinity logits.
It is not tested as far as I know.
I manually confirmed the behavior but I'm not convinced that it always works well.

2. Is it better that resolve the warning caused by infinity logits?
When I manually confirmed the behavior of infinity logits, I got a RuntimeWarning message.
It might be better that resolve the message since the code explicitly handles infinity logits as p=1.

Both of them might be included this PR or resolved in new PRs.

I confirmed the behavior of logit containing an infinity value like following.
```
>>> import chainer as ch             
>>> import numpy as np                                                                      
>>> from chainer.distributions.bernoulli import Bernoulli
>>> logit = ch.Variable(np.array([0, np.inf]))                                              
>>> x = ch.Variable(np.array([1., 1.]))       
>>> b = Bernoulli(logit=logit, binary_check=True)                                           
>>> logprob = b.log_prob(x)
/src/chainer/distributions/bernoulli.py:27: RuntimeWarning: invalid value encountered in multiply
  y = logit * (x - 1) - xp.log(xp.exp(-logit) + 1)                                          
>>> logprob            
variable([-0.69314718,  0.        ])          
>>> sum(logprob).backward()                   
>>> logit.grad         
array([0.5, nan])  
```
The result above is same as Chainer 6.0.0rc1.

**Comment by toslunar** [comment]
>     1. Is it better that tests for infinity logits exist?
>        [a0256a6](https://github.com/chainer/chainer/commit/a0256a672e30a6866de2e23c9370953975bf6724) changes the implementation to handle infinity logits.
>        It is not tested as far as I know.
>        I manually confirmed the behavior but I'm not convinced that it always works well.

I'd like to have tests that the backward just runs, while I don't think it's mandatory to check values of grads in the PR because the *forward* tests are weak and the PR fixes the backward.  The backward tests may hard-code the expected grads since `gradient_check.numerical_grad` doesn't support infinities.

On forward tests, infinity logits are tested via the testing parameter `self.extreme_values=True` of `TestBernoulli`, where several issues have been left because of low-priority.

- For non-binary inputs there should be 8 test methods: {`prob`, `log_prob`} × {`binary_check=True`, `binary_check=False`} × {`cpu`, `gpu`}, but there are 4.
- Value tests for non-binary inputs are missing.  Such tests could be added to `check_log_prob_binary_check` and `check_prob_binary_check`, whose name would be renamed to something like `*_non_binary`.

**Comment by seiyab** [comment]
@toslunar 
Thank you for your response.
Let me confirm your indications.

Does
> I'd like to have tests that the backward just runs...

mean
> It seems better to have tests that confirm backward method doesn't raise error when logit have infinity value.

rather than
> It seems better to have tests that hard-code the expected grads for logits of infinity value.

or
> The existing tests that checks grads for finite values are enough.

?

About forward, now I think there is no need to worry about it in this PR.
I'll send another PR if I want to fix them.

**Comment by toslunar** [comment]
@seiyab 
> Does
> 
> > I'd like to have tests that the backward just runs...
> 
> mean
> 
> > It seems better to have tests that confirm backward method doesn't raise error when logit have infinity value.

Yes.  Thanks for clarifying it.

**Comment by seiyab** [comment]
@toslunar 
I added a test.

File: tests/chainer_tests/distributions_tests/test_bernoulli.py [hunk 2]
```diff
@@ -156,5 +163,16 @@ def test_double_backward_gpu(self):
             cuda.to_gpu(self.logit), cuda.to_gpu(self.x),
             cuda.to_gpu(self.gy), cuda.to_gpu(self.ggx))
 
+    def test_backward_where_logit_has_infinite_values(self):
+        self.logit += numpy.inf
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_bernoulli.py]: For the readability,
```suggestion
        self.logit[...] = numpy.inf
```
looks better to me.

File: tests/chainer_tests/distributions_tests/test_bernoulli.py [hunk 3]
```diff
@@ -156,5 +163,16 @@ def test_double_backward_gpu(self):
...
+        with numpy.errstate(invalid='ignore'):
+            log_prob = distributions.bernoulli._bernoulli_log_prob(
+                self.logit, self.x)
+
+        log_prob.backward()
+
+        # just confirm that the backward method runs without raising error.
+        assert True
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_bernoulli.py]: How does the line confirm it?

File: tests/chainer_tests/distributions_tests/test_bernoulli.py (hunk 2 above)
Comment (seiyab) [reply tests/chainer_tests/distributions_tests/test_bernoulli.py]: Thank you.
I'll fix it.

File: tests/chainer_tests/distributions_tests/test_bernoulli.py (hunk 3 above)
Comment (seiyab) [reply tests/chainer_tests/distributions_tests/test_bernoulli.py]: I'm sorry for misleading code.

Is the following without assert statement clearer?
```
# just confirm that the backward method runs without raising error.
log_prob.backward()
```

What I wanted to express is following.
> This test method just confirms that the backward method runs without raising error.
> So assert statement is always passed.
> `assert True` explicitly expresses there is nothing to assert.

Comment (toslunar) [reply tests/chainer_tests/distributions_tests/test_bernoulli.py]: In the repository `assert True` is not used.  How about replacing the last 4 lines with the following?
```
        # just confirm that the backward method runs without raising error.
        log_prob.backward()
```

Comment (seiyab) [reply tests/chainer_tests/distributions_tests/test_bernoulli.py]: I adopted that. Thank you.

File: tests/chainer_tests/distributions_tests/test_bernoulli.py (hunk 2 above)
Comment (seiyab) [reply tests/chainer_tests/distributions_tests/test_bernoulli.py]: I fixed it.

**Comment by toslunar** [comment]
Jenkins, test this please.
